## Importing Libraries

In [2]:
import pandas as pd
import numpy as np
import os

## Importing Data

In [5]:
path = r'C:\Users\ariel\OneDrive\Desktop\School work\08-2020 Instacart Basket Analysis'

In [7]:
path

'C:\\Users\\ariel\\OneDrive\\Desktop\\School work\\08-2020 Instacart Basket Analysis'

In [53]:
#Importing ords_prods_merge DataFrame
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [55]:
#Creating a subset of the data
df = df_ords_prods_merge[:1000000]

## Adding Price Label

In [57]:
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [59]:
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\ariel\AppData\Local\Temp\ipykernel_10248\2848743372.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [60]:
#Checking dimensions
df.shape

(1000000, 16)

In [63]:
#Checking product counts
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    652638
Low-range product    338018
High range             9344
Name: count, dtype: int64

In [65]:
df['prices'].max()

24.5

## If-Statements with the loc() Function

In [67]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\ariel\AppData\Local\Temp\ipykernel_10248\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [69]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [71]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [73]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     652638
Low-range product     338018
High-range product      9344
Name: count, dtype: int64

## Using If-Statements on full data frame

In [76]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [78]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [80]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [82]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

## If-Statements with For-Loops

In [89]:
#Finding busiest and slowest days of the week
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

Day 0 is Saturday. Saturday is the busiest day of the week, with the most amount of orders on that day. Wednesday (4) is the slowest day of the week, with the lowest amount of orders on that day.

In [94]:
#Creating a new column portraying the busiest day, least busy, and regularly busy
result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [96]:
result

['Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Bus

In [98]:
#Combining results with the ords_prods_merge dataframe
df_ords_prods_merge['busiest_day'] = result

In [102]:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

In [104]:
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  orders_day_of_week  \
0   3139998      138    prior            28                   6   
1   1977647      138    prior            30                   6   
2    389851      709    prior             2                   0   
3    652770      764    prior             1                   3   
4   1813452      764    prior             3                   4   

   order_hour_of_day  days_since_last_order  add_to_cart_order  reordered  \
0                 11                   3.00                  5          0   
1                 17                  20.00                  1          1   
2                 21                   6.00                 20          0   
3                 13                  11.11                 10          0   
4                 17                   9.00                 11          1   

  _merge    price_range_loc     busiest_day  
0   both  Mid-range product  Regularly busy  
1   both  Mid-range product  Regularly busy  
2   both  Mid-range product     Busiest day  
3   both  Mid-range product  Regularly busy  
4   both  Mid-range product      Least busy

Regularly busy days: Days 1,2,3,5,6. Total orders on regular days: 22,416,875.

Busiest day: Day 0. Total orders on busiest day: 6,204,182.

Least busy day: Day 4. Total orders on least busy day: 3,783,802.

## Task 4.7 #2. New Column containing "Busiest_Days"

In [131]:
#Create a new column changing labels in "Busiest Day" column to "Busiest Days". With two busiest days and two slowest days.
results = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value ==1:
    results.append("Busiest days")
  elif value == 4 or value==3:
    results.append("Least busy days")
  else:
    results.append("Regularly busy")

In [133]:
results

['Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Least busy days',
 'Least busy days',
 'Busiest days',
 'Regularly busy',
 'Least busy days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Least busy days',
 'Least busy days',
 'Regularly busy',
 'Least busy days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Least busy days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Least busy days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Least busy days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Least busy days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Least busy days',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',


In [139]:
#Combining results with the ords_prods_merge dataframe
df_ords_prods_merge['busiest_days'] = results

In [141]:
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  orders_day_of_week  \
0   3139998      138    prior            28                   6   
1   1977647      138    prior            30                   6   
2    389851      709    prior             2                   0   
3    652770      764    prior             1                   3   
4   1813452      764    prior             3                   4   

   order_hour_of_day  days_since_last_order  add_to_cart_order  reordered  \
0                 11                   3.00                  5          0   
1                 17                  20.00                  1          1   
2                 21                   6.00                 20          0   
3                 13                  11.11                 10          0   
4                 17                   9.00                 11          1   

  _merge    price_range_loc     busiest_day     busiest_days  
0   both  Mid-range product  Regularly busy   Regularly busy  
1   both  Mid-range product  Regularly busy   Regularly busy  
2   both  Mid-range product     Busiest day     Busiest days  
3   both  Mid-range product  Regularly busy  Least busy days  
4   both  Mid-range product      Least busy  Least busy days

## #3.

In [147]:
#Checking frequency of new column
df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy     12916111
Busiest days       11864412
Least busy days     7624336
Name: count, dtype: int64

Total order numbers are : 32,404,859. This is the same as the busiest day column. They still follow the same order with "Regularly busiest days" accumulating the most orders together.

 According to the new "Busiest Days" column, now Saturday and Sunday are among the busiest days. Tuesday and Wednesday are the slowest days. Monday, Thursday, and Friday are among the regularly busy days.

## #4. New Column Containing "Busiest_period_of_day"

In [154]:
#Finding busiest hours of the day
df_ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [156]:
result_2 = []

for value in df_ords_prods_merge["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result_2.append("Most orders")
  elif value in [23, 6, 0, 1, 5, 2, 4, 3]:
    result_2.append("Fewest orders")
  else:
    result_2.append("Average orders")

In [158]:
result_2

['Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Average orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most ord

In [160]:
#Incorporating new column into dataframe
df_ords_prods_merge['busiest_period_of_day'] = result_2

## #5.

In [167]:
#Checking frequency of new column
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

## Exporting Dataframe

In [170]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))